R/component-LSM.R

```diff
@@ -0,0 +1,292 @@
+      self$latent.space.pos.v <- latent.space.pos.v
+      self$latent.space.pos.v.ab <- latent.space.pos.v.ab
+      self$latent.space.target <- latent.space.target
+      self$inverted.model <- inverted.model
+      self$tune <- tune
+    },
+
+    create.component = function(n.nodes, edge.list, node.names) {
+      return(LSMComponent$new(n.nodes, edge.list, node.names, self))
+    }
+  )
+)
+
+# Component Class Definition --------------------------------------------
+LSMComponent <- R6Class(
+  classname = "LSMComponent",
+  inherit = BaseComponent,
+  public = list(
+    dimension = 2,
+    latent.space.pos = NULL,
+    latent.space.pos.m = 0,
+    latent.space.pos.v = 100,
+    latent.space.pos.v.ab = c(0.001, 0.001),
+    latent.space.target = NULL,
+    inverted.model = FALSE,
+    tune = 0.1,
+
+    # TODO: What else should be initialized?
+    initialize = function(n.nodes, edge.list, node.names, params) {
+      self$dimension <- dimension
+      self$latent.space.pos <- params$latent.space.pos
+      self$latent.space.pos.m <- params$latent.space.pos.m
+      self$latent.space.pos.v <- params$latent.space.pos.v
+      self$latent.space.pos.v.ab <- params$latent.space.pos.v.ab
+      self$latent.space.target <- params$latent.space.target
+      self$inverted.model <- params$inverted.model
+      self$tune <- params$tune
+      
+      private$n.nodes <- n.nodes
+      private$edge.list <- edge.list
+      private$node.names <- node.names
+      private$n.edges <- nrow(edge.list) # QUESTION: Need this?
+      private$mult.factor <- 2 * inverted.model - 1
+    },
+
+    random.start = function() {
+      "Generates a random latent positions to initiate the MCMC sampling. Places
+      Them in the correct object field."
+      n <- self$dimension * self$n.nodes
+      inits <- rnorm(n, 0, sqrt(self$latent.space.pos.v))
+      self$latent.space.pos <- matrix(inits, nrow = self$n.nodes)
+      self$latent.space.target <- self$latent.space.pos
+    },
+    
+    log.likelihood = function(parameters = pieces(),
+                              edges = seq_len(nrow(edge.list))) {
+      "Computes log-likelihood of outcome given with the mean being predicted by
+    a regression on the latent-space distances."
+      meanpart <- value.ext(parameters, edges)
+      sum(dnorm(outcome[edges], meanpart, sqrt(residual.variance), log = TRUE))
+    },
+    
+    draw = function(verbose = 0, mh.tune = tune, langevin.mode = FALSE) {
+      "Computes every step for each MCMC draw. MH step for latent space
+      positions, Gibbs update for variance."
+      # TODO: Update to be broken into private methods
+      # NOTES: This could be more self-documenting by breaking it into more
+      #        functions? It's also very slow looking. Not exploiting
+      #        vectorization.
+      
+      # dimension of latent space positions
+      # (why do we need a field called "dimension" then??)
+      lsdim <- self$dimension
```

Comment:
I agree, we should just refer to self$dimension.

I think there are a lot of parts of the original code that unnecessarily defined local parameters.